bin/as-recon-discover-services.tengo

```diff
@@ -0,0 +1,195 @@
+#! /usr/bin/env arsenic
+
+// arsenic := import("arsenic")
+cobra := import("cobra")
+os := import("os")
+os2 := import("os2")
+filepath := import("filepath")
+check_err := import("check_err")
+nmap := import("nmap")
+script := import("script")
+log := import("log")
+
+topPorts := 0
+targets := []
+
+setupHost := func(target) {
+  baseDir := ""
+  if filepath.dir_exists("hosts") {
+    baseDir = filepath.join("hosts", target)
+  }
+
+  err := os.mkdir_all(filepath.join(baseDir, "loot"), 0755)
+  check_err(err)
+
+  err = os.mkdir_all(filepath.join(baseDir, "recon"), 0755)
+  check_err(err)
+
+  return baseDir
+}
+
+defaultScanner := func(target) {
+  return nmap.scanner().
+    stylesheet("/static/nmap.xsl").
+    open().
+    Pn().
+    n().
+    system_dns().
+    targets(target)
+}
+
+rootCmd := cobra.root_cmd("as-recon-discover-services.tengo", "Run nmap scans on targets")
+rootCmd.enable_completion()
+
+rootCmd.persistent_flags.string_slicep("targets", "t", [], "The target to scan. Can be used multiple times to specify multiple targets")
+rootCmd.persistent_flags.stringp("target-file", "f", "", "The file containing the targets to scan.")
+rootCmd.persistent_flags.int("top", 0, "Scan # most common ports")
+
+rootCmd.set_persistent_pre_run(func(cmd, args) {
+  targets = cmd.flags.get_string_slice("targets")
+  file := cmd.flags.get_string("target-file")
+
+  if file != "" {
+    lines := os2.read_file_lines(file)
+    check_err(lines)
+
+    for line in lines {
+      targets = append(targets, line)
+    }
+  }
+
+  for line in os2.read_stdin() {
+    targets = append(targets, line)
+  }
+
+  if len(targets) == 0 {
+    script.stop("No targets specified")
+  }
+
+  topPorts = cmd.flags.get_int("top")
+})
+
+// TCP Command
+
+tcpCmd := cobra.cmd("tcp", "Run TCP scans")
+rootCmd.add_command(tcpCmd)
+
+tcpCmd.set_run(func(cmd, args){
+  topScan := func(target) {
+    log.msg(format("Scanning %s: top %d TCP ports", target, topPorts))
+
+    scanner := defaultScanner(target).
+      privileged().
+      oA(filepath.from_slash(format("recon/nmap-top-%d-tcp", topPorts))).
+      host_timeout("30m").
+      A().
+      top_ports(topPorts)
+    
+    run := scanner.run_direct()
+    check_err(run)
+  }
+
+  quickScan := func(target) {
+    scanner := defaultScanner(target).
+      T3().
+      host_timeout("30m").
```

Comment:
we should make this an arg and default it to a configurable value

Reply:
Fully agree with that one. Was already thinking about that today since I ran into an instance where I knew the host was up and had ports open but the full TCP scan timed out because of the `30m` timeout we set